pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/PolicyHierarchyValue.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.policies.data;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import lombok.Getter;
+
+/**
+ * Policy value holder for different hierarchy level.
+ * Currently, we have three hierarchy with priority : topic > namespace > broker.
+ */
+public class PolicyHierarchyValue<T> {
+    private static final AtomicReferenceFieldUpdater<PolicyHierarchyValue, Object> VALUE_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(PolicyHierarchyValue.class, Object.class, "value");
+
+    private volatile T brokerValue;
+
+    @VisibleForTesting
+    @Getter
+    private volatile T namespaceValue;
+
+    private volatile T topicValue;
+
+    private volatile T value;
+
+    public PolicyHierarchyValue() {
+    }
+
+    public void updateBrokerValue(T brokerValue) {
+        this.brokerValue = brokerValue;
+        updateValue();
+    }
+
+    public void updateNamespaceValue(T namespaceValue) {
+        this.namespaceValue = namespaceValue;
+        updateValue();
+    }
+
+    public void updateTopicValue(T topicValue) {
+        this.topicValue = topicValue;
+        updateValue();
+    }
+
+    private void updateValue() {
+        VALUE_UPDATER.updateAndGet(this, (preValue) -> {
```

Comment:
We might need to handle the delete policies? Such as the topic level policy being removed, we should use the namespace level policy if present.

Reply:
Yes, in this case, `updateTopicValue(null)` should be called, and the `value` will be updated to `namespaceValue` if it's presented.

Reply:
Checked with `org.apache.pulsar.broker.admin.v2.PersistentTopics#removeMaxSubscriptionsPerTopic`, `internalSetMaxSubscriptionsPerTopic(null)` is called to update the policy.